contentcuration/contentcuration/utils/pagination.py

```diff
@@ -1,84 +1,121 @@
+    def __init__(self, object_list, *args, **kwargs):
+        if not isinstance(object_list, QuerySet):
+            raise TypeError("ValuesViewsetPaginator is only intended for use with Querysets")
+        self.queryset = object_list
+        object_list = object_list.values_list("pk", flat=True).distinct()
+        return super(ValuesViewsetPaginator, self).__init__(object_list, *args, **kwargs)
+
+    def _get_page(self, object_list, *args, **kwargs):
+        pks = list(object_list)
+        return ValuesPage(self.queryset.filter(pk__in=pks), *args, **kwargs)
+
+
+class CachedValuesViewsetPaginator(ValuesViewsetPaginator):
+    @cached_property
+    def count(self):
+        try:
+            query_string = str(self.object_list.query).encode("utf8")
             cache_key = (
                 "query-count:"
-                + hashlib.md5(str(queryset.query).encode("utf8")).hexdigest()
+                + hashlib.md5(query_string).hexdigest()
             )
             value = cache.get(cache_key)
             if value is None:
-                value = queryset.values("id").count()
+                value = super(CachedValuesViewsetPaginator, self).count
                 cache.set(cache_key, value, 300)  # save the count for 5 minutes
-            return value
+        except EmptyResultSet:
+            # If the query is an empty result set, then this error will be raised by
+            # Django - this happens, for example when doing a pk__in=[] query
+            # In this case, we know the value is just 0!
+            value = 0
+        return value
+
+
+class ValuesViewsetPageNumberPagination(PageNumberPagination):
+    django_paginator_class = ValuesViewsetPaginator
 
+    def paginate_queryset(self, queryset, request, view=None):
+        """
+        Paginate a queryset if required, either returning a
+        page object, or `None` if pagination is not configured for this view.
+        """
         page_size = self.get_page_size(request)
         if not page_size:
             return None
 
         paginator = self.django_paginator_class(queryset, page_size)
-        paginator.count = cached_count(queryset)
         page_number = request.query_params.get(self.page_query_param, 1)
         if page_number in self.last_page_strings:
             page_number = paginator.num_pages
 
         try:
             self.page = paginator.page(page_number)
         except InvalidPage as exc:
-            msg = self.invalid_page_message.format(page_number=page_number, message=exc)
+            msg = self.invalid_page_message.format(
+                page_number=page_number, message=str(exc)
+            )
             raise NotFound(msg)
 
         if paginator.num_pages > 1 and self.template is not None:
             # The browsable API should display pagination controls.
             self.display_page_controls = True
 
         self.request = request
-        return list(self.page)
+        return self.page.queryset
```

Comment:
Not really your change, but this looks mostly taken from the [rest framework class](https://github.com/encode/django-rest-framework/blob/master/rest_framework/pagination.py). A comment noting the difference here would be nice

Reply:
Yup, can do, and yes it is.